{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "626c3f1c",
   "metadata": {},
   "source": [
    "## Javascript promises and Async programming\n",
    "\n",
    "### Fetching data\n",
    "* one way to make sure the order of function calls is to use pyramid structure of functions, as shown below:\n",
    "```javascript\n",
    "    a(\"test\", (err, aResult) => {\n",
    "      b(aResult, (err, bResult) => {\n",
    "        c(bResult, (err, cResult) => {\n",
    "          d(cResult);\n",
    "        });\n",
    "      });\n",
    "    });\n",
    "\n",
    "```\n",
    "  + a executes before b, which executes before c, which executes before d \n",
    "  + dirty code\n",
    "  + error handling difficult\n",
    "* promise\n",
    "  + an object that represents the eventual completion (or failure) of an asynchronous operation, and its resulting value\n",
    "  + readable asynchronous code\n",
    "* promise states\n",
    "  + pending\n",
    "  + fulfilled (return a single value)\n",
    "  + rejected (when a asynchronous call has failed)\n",
    "    + it returns with the reason why it failed similar to the catch function in try/catch block\n",
    "* code example\n",
    "  + axios send HTTP GET request to the url, when sends back results as a Promise\n",
    "  + the get function is chained to then function which will be called once the get function get the Promise object with request fulfilled\n",
    "  + in then function, we first desctruct the return JSON object and get data property, then use the retrieved data object to set the text in result box\n",
    "  + if the promise is rejected, then the error will be sent to catch function (the only object sent to catch is the error)\n",
    "  + note that a fulfilled promise will send the entire result obj to then function, that is why we need to destructure the data property, but the rejected promise only sends the error to catch function\n",
    "  \n",
    "```javascript\n",
    "\n",
    "export function get() {\n",
    "    axios.get(\"http://localhost:3000/orders/1\")\n",
    "    .then(({data}) => {\n",
    "        setText(JSON.stringify(data));\n",
    "    })\n",
    "    .catch(err => setText(err)); \n",
    "}\n",
    "```\n",
    "\n",
    "### chaining Promises\n",
    "* both then and catch functions return Promises (Promises return Promises). This allows us to chain Promises together\n",
    "* in the following code example\n",
    "  + the first then function set the text, and then return \"Pluralsight\" (you can return anything)\n",
    "  + this string is wrapped as a Promise object and returned by then\n",
    "  + the returned Promise object is then console logged\n",
    "```javascript\n",
    "  axios.get(\"http://localhost:3000/orders/1\").then(({ data }) => {\n",
    "      setText(JSON.stringify(data));\n",
    "      return \"Pluralsight\";\n",
    "  })\n",
    "  .then(result => console.log(result))\n",
    "```\n",
    "\n",
    "* let's chain two HTTP GET requests together (remember to add return in the first then clause\n",
    "  + the catch function will catch any errors in either of the two then functions\n",
    "```javascript\n",
    "  axios.get(\"http://localhost:3000/orders/1\").then(({ data }) => {\n",
    "      return axios.get(`http://localhost:3000/addresses/${data.shippingAddress}`\n",
    "     })\n",
    "  .then(({data}) => {\n",
    "    setText(`City: ${data.city}`);\n",
    "  })\n",
    "  .catch(err => setText(err));\n",
    "}\n",
    "```\n",
    "\n",
    "* if we have two requests, how to make sure the errors specific to each then function will be caught separately, rather than being caught by the second catch function?\n",
    "  + we can put a catch function for each of the then function\n",
    "  + here the second catch function will catch the error from both the first catch and second then functions\n",
    "```javascript\n",
    "  axios.get(\"http://localhost:3000/orders/1\").then(({ data }) => {\n",
    "      return axios.get(`http://localhost:3000/addresses/${data.shippingAddress}`\n",
    "     })\n",
    "  .catch(err => {\n",
    "      setText(err);\n",
    "      return {data:{}};\n",
    "  })\n",
    "  .then(({data}) => {\n",
    "    setText(`City: ${data.city}`);\n",
    "  })\n",
    "  .catch(err => setText(err));\n",
    "}\n",
    "```  \n",
    "* run code after either then and catch (in this example, to hide the waiting sign\n",
    "```javascript\n",
    "  axios.get(\"http://localhost:3000/orders/1\").then(({ data }) => {\n",
    "      return axios.get(`http://localhost:3000/addresses/${data.shippingAddress}`\n",
    "     })\n",
    "  .then(({data}) => {\n",
    "    setText(`City: ${data.city}`);\n",
    "  })\n",
    " .catch(err => setText(err))\n",
    " .finally(() => {\n",
    "      setTimeout(() => {\n",
    "          hideWaiting();\n",
    "      }, 1500);\n",
    "  });\n",
    "}\n",
    "```  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c74c5b0",
   "metadata": {},
   "source": [
    "### Creating promises\n",
    "* a pending promise is a promise that has not yet settled\n",
    "* Promise\n",
    "  + object that represents the eventual completion (or failure) of an asynchronous operation, and its resulting value\n",
    "* how to create a Promise\n",
    "  + use new Promise, with a parameter called resolve\n",
    "  + define the function body, and in resolve(), define the object to return when request is fulfilled\n",
    "  + consume the Promise by then() function and you will see the \"timeout!\" text is the object returned from resolve\n",
    "```javascript\n",
    "\n",
    "export function timeout() {\n",
    "    \n",
    "    const wait = new Promise((resolve) => {\n",
    "        setTimeout(() => {\n",
    "            resolve(\"Timeout!\");\n",
    "        }, 1500);\n",
    "    });\n",
    "\n",
    "    wait.then(text => setText(text));\n",
    "}\n",
    "\n",
    "```\n",
    "\n",
    "* what about functions that will be invoked multiple times by setInterval?\n",
    "  + setInterval fires multiple times while setTimeout fires only once\n",
    "  + you will only see the first execution of the function fired. Once a Promise is settled, it is done. The following firing of the resolve() function will not be executed, although console.log is successfully fired multiple times\n",
    "  + the function defined in Promise still keeps running, but not the resolve() function\n",
    "```javascript\n",
    "\n",
    "export function interval() {\n",
    "    let counter = 0;\n",
    "    const wait = new Promise((resolve) => {\n",
    "        setInterval(() => {\n",
    "            console.log(\"INTERVAL\");\n",
    "            resolve(`Timeout! ${++counter}`);\n",
    "        }, 1500);\n",
    "    });\n",
    "\n",
    "    wait.then(text => setText(text));\n",
    "    .finally(() => appendText(` -- Done ${counter}`));\n",
    "}\n",
    "\n",
    "```  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b4388f",
   "metadata": {},
   "source": [
    "   \n",
    "* add both resolve and reject to handle fulfill and reject states\n",
    "  + when creating a Promise\n",
    "    + for fulfill state, return the object/data using resolve()\n",
    "    + for reject state, return the error message/object uisng reject()\n",
    "  + when consuming the Promise\n",
    "    + use then and its parameter representing the object/data returned from resolve()\n",
    "    + use catch and its parameter representing the error message/object returned from reject()\n",
    "  \n",
    "```javascript\n",
    "export function xhr() {\n",
    "    let request = new Promise((resolve, reject) => {\n",
    "        let xhr = new XMLHttpRequest();\n",
    "        xhr.open(\"GET\", \"http://localhost:3000/users/7\");\n",
    "        xhr.onload = () => {\n",
    "            if (xhr.status === 200) {\n",
    "                resolve(xhr.responseText);\n",
    "            } else {\n",
    "                reject(xhr.statusText);\n",
    "            }\n",
    "        }\n",
    "        xhr.onerror = () => reject(\"Request Failed\");\n",
    "        xhr.send();\n",
    "    });\n",
    "    \n",
    "    request.then(result=> setText(result))\n",
    "    .catch(reason => setText(reasson));\n",
    "}\n",
    "```\n",
    "\n",
    "* waiting for all promises to resolve using Promise.all()\n",
    "  + call HTTP GET by axios\n",
    "  + wait until all API asynchronous calls are returned by Promise.all(), and define the order to process these Promises\n",
    "  + in then() define cat, stat and type Promise objects, and append the content to text by appendText\n",
    "  + Promise.all() wait return when \n",
    "    + either all Promises are fulfilled\n",
    "    + or one of them is rejected (the first rejected Promise will cause the function to return)\n",
    "      + the returned rejected obj will be handled by catch()\n",
    "```javascript\n",
    "    export function allPromises() {\n",
    "        let category = axios.get(\"http://localhost:3000/tiemCategories\");\n",
    "        let statuses = axios.get(\"http://localhost:3000/orderStatuses\");\n",
    "        let userTypes = axios.get(\"http://localhost:3000/userTypes\");\n",
    "\n",
    "        Promise.all([category, statuses, userTypes])\n",
    "        .then(([cat, stat, type]) => {\n",
    "            setText(\"\");\n",
    "\n",
    "            appendText(JSON.stringify(cat.data));\n",
    "            appendText(JSON.stringify(stat.data));\n",
    "            appendText(JSON.stringify(type.data));\n",
    "        })\n",
    "        .catch(err => setText(err));\n",
    "    }\n",
    "```\n",
    "\n",
    "* what if we don't need to wait all the Promises to be resolved, if some of them are rejected, we can still proceed? We use Promises.allSettled\n",
    "  + Promises.allSettled returns two JS object, one is resolved, the other is rejected\n",
    "    + in resolved there are two properties: status: fulfilled and value: {}\n",
    "    + in rejected, there are two properties: status: rejected and reason: {}\n",
    "    + since both resolved and rejected are returned in one object, we only need to use .then to consume, and no need to use catch, but is recommended to catch other errors\n",
    "    \n",
    "    ```javascript\n",
    "        // resolved.js \n",
    "        {\n",
    "            status: \"fulfilled\",\n",
    "                value: {}\n",
    "        }\n",
    "\n",
    "        // rejected.js\n",
    "        {\n",
    "            status: \"rejected\",\n",
    "                reason: {}\n",
    "        }\n",
    "\n",
    "    ```\n",
    "  + code example of using allSettled\n",
    "  ```javascript\n",
    "    export function allSettled() {\n",
    "        let category = axios.get(\"http://localhost:3000/tiemCategories\");\n",
    "        let statuses = axios.get(\"http://localhost:3000/orderStatuses\");\n",
    "        let userTypes = axios.get(\"http://localhost:3000/userTypes\");\n",
    "\n",
    "        Promise.allSettled([category, statuses, userTypes])\n",
    "        .then((values) => {\n",
    "            let results = values.map(v => {\n",
    "                if(v.status === 'fulfilled'){\n",
    "                    return `FULFILLED: ${JSON.stringify(v.value.data[0])} `\n",
    "                    }\n",
    "                    \n",
    "                return `REJECTED: ${v.reason.message} `;\n",
    "                });\n",
    "\n",
    "            setText(results);\n",
    "         })\n",
    "        .catch(err => setText(err));\n",
    "    }\n",
    "```\n",
    "\n",
    "* racing Promises\n",
    "  + if you have several API calls for the same information, but just need to get the fasted returned results and don't need to wait for other reponses, you can use Promise.race() function\n",
    "  + race() stops when the first response is settled, which may be resolved or rejected. Even if the later response returns the resolved data, the function still directly goes to catch block, if the first reponse received is a rejected result.\n",
    "  ```javascript \n",
    "  export function race() {\n",
    "      let users = axios.get(\"http://localhost:3000/users\");\n",
    "      let backup = axios.get(\"http://localhost:3001/users\");\n",
    "      \n",
    "      Promise.race([users, backup])\n",
    "      .then(users => setText(JSON.stringify(users.data)))\n",
    "      .catch(reason => setText(reason));\n",
    "  } "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ded77f3e",
   "metadata": {},
   "source": [
    "### Async/Await\n",
    "* Async/await is syntactic sugar\n",
    "  + async\n",
    "    + designate that a function is asynchronous\n",
    "    + the function returns a Promise that wrap whatever you return in the function\n",
    "    \n",
    "    ```javascript\n",
    "\n",
    "    // async\n",
    "    async function getNames() {\n",
    "        return [];\n",
    "    }\n",
    "\n",
    "    const getNames = async () => {\n",
    "        return [];\n",
    "    }\n",
    "```\n",
    "\n",
    "  + await\n",
    "    + pauses the execution of an asynchronous function while it waits for Promise to fulfill\n",
    "    + can only be used inside an async function\n",
    "    + only blocks the current function\n",
    "    + it does not block calling function\n",
    "    + in the following code example\n",
    "      + await will pause the code inside getNames to wait until someFunc() returns, and therefore, delay the execution of doSomethingElse()\n",
    "      + await will not impact getAddresses()\n",
    "    \n",
    "    ```javascript\n",
    "\n",
    "        // await code example\n",
    "\n",
    "        const getNames = async () => {\n",
    "            await someFunc();\n",
    "            doSomethingElse();\n",
    "        }\n",
    "\n",
    "        getNames();\n",
    "        getAddresses();\n",
    "\n",
    "    ```\n",
    "  \n",
    "* Use the standard try/catch to handle both synchronous and asynchronous code\n",
    "\n",
    "```javascript\n",
    "\n",
    "    export async function get() {\n",
    "        try {\n",
    "        const {data} = await axios.get(\"http://localhost:3000/orders/1\");\n",
    "        setText(JSON.stringify(data));\n",
    "        } catch (error) {\n",
    "            setText(error);\n",
    "        }        \n",
    "    }\n",
    "\n",
    "    export async function get() {\n",
    "        axios.get(\"http://localhost:3000/orders/1\")\n",
    "        .then(({data}) => {\n",
    "        setText(JSON.stringify(data))\n",
    "    })\n",
    "     .catch(err => setText(err));\n",
    "    }   \n",
    "\n",
    "```\n",
    "\n",
    "* chain async/await\n",
    "\n",
    "```javascript\n",
    "\n",
    "    export async function chain() {\n",
    "        const {data} = await axios.get(\"http://localhost:3000/orders/1\");\n",
    "\n",
    "        // destructure data and assign it to address\n",
    "        const {data: address} = await axios.get(`http://localhost:3000/addresses/${data.shippingAddress}`);\n",
    "\n",
    "        setText(`City: ${JSON.stringify(address.city)}`);\n",
    "    }\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c52daf8",
   "metadata": {},
   "source": [
    "* await concurrent requests\n",
    "  + this is done by first defining the functions as function expressions, and then call await. The two functions will be called at the same time\n",
    "  + the following appendText calls will wait until both await functions return\n",
    "```javascript\n",
    "\n",
    "    export async function concurrent() {\n",
    "        // define two functions first as function expressions\n",
    "        const orderStatus = axios.get(\"http://localhost:3000/orderStatus\");\n",
    "        const orders = axios.get(\"http://localhost:3000/orders\");\n",
    "\n",
    "        setText(\"\");\n",
    "        \n",
    "        const {data: statuses} = await orderStatus;\n",
    "        const {data: order} = await orders;\n",
    "        \n",
    "        appendText(JSON.stringify(statuses));\n",
    "        appendText(JSON.stringify(order[0]));\n",
    "        \n",
    "    }\n",
    "\n",
    "```\n",
    "\n",
    "* awaiting parallel calls\n",
    "  + use the await decorated Promise.all, and list the async anonymous functions IIFEs to execute each sync calls simultaneously\n",
    "   + immediately invoked function expression\n",
    "```javascript\n",
    "\n",
    "    export async function parallel() {\n",
    "        setText(\"\");\n",
    "        \n",
    "        await Promise.all([\n",
    "            (async () => {\n",
    "                const {data} = await axios.get(\"http://localhost:3000/orderStatus\");\n",
    "                appendText(JSON.stringify(data));\n",
    "            })(),\n",
    "            (async () => {\n",
    "                const {data} = await axios.get(\"http://localhost:3000/orders\");\n",
    "                appendText(JSON.stringify(data));\n",
    "            })()\n",
    "            \n",
    "        ]);\n",
    "        \n",
    "    }\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f855069",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
